### PR TITLE
DP-22263  Implement responsive images for promotional page images in key message sections

### PR DESCRIPTION
**Pull request by Javi-er**
Any PRs being created needs a changelog.txt file before being merged into dev. See: [Change Log Instructions](https://github.com/massgov/mayflower/blob/develop/docs/for-developers/changelog-instructions.md)

## Description
Adjusting "key message" component to use an element instead of a background image in order to support responsive images through the picture element. 
Also I adjusted the marketing campaign page to use a responsive image as well.

## Related Issue / Ticket

- [JIRA issue](https://massgov.atlassian.net/browse/DP-22263)

## Steps to Test

1. Checkout branch on local mayflower.
2. Start patternlab with `rush start:patternlab`
3. Visit the marketing page and inspect the key message image.

## Screenshots
Use something like [licecap](http://www.cockos.com/licecap/) to capture gifs to demonstrate behaviors.

## Additional Notes:

Anything else to add?

#### Impacted Areas in Application

*

#### @TODO

*

### CODE REVIEW

File: changelogs/DP-22263.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+# ChangeType [enter a valid option, see __CHANGE TYPES___ e.g. Added]:
+#  - project: Enter one or multiple (comma separated) valid project prefix(es) - see ___PROJECT___ e.g. React, Patternlab
+#    component: Enter one or multiple (comma separated) valid component name(s) - see ___COMPONENT___ e.g. Color
+#    description: Describe the change and add the PR number. see ___DESCRIPTION___ e.g. Adds apples to apple trees for admin apple pickers (#PR)
+#    issue: Add a Jira ticket or issue number, e.g. DP-12345 or 124
+#    impact: Enter a valid impact, e.g. Minor
+
+#___EXAMPLE___
+Changed:
+  - project: React, Patternlab
```
Comment (clairesunstudio) [inline changelogs/DP-22263.yml]:
```suggestion
  - project: Patternlab
```

File: changelogs/DP-22263.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    component: Key Message
+    description: Adjusted to support reponsive images (picture)
+    issue: DP-22263
+    impact: Minor
+  - project: React, Patternlab
```
Comment (clairesunstudio) [inline changelogs/DP-22263.yml]:
```suggestion
  - project: Patternlab
```

File: changelogs/DP-22263.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+#  - project: Enter one or multiple (comma separated) valid project prefix(es) - see ___PROJECT___ e.g. React, Patternlab
+#    component: Enter one or multiple (comma separated) valid component name(s) - see ___COMPONENT___ e.g. Color
+#    description: Describe the change and add the PR number. see ___DESCRIPTION___ e.g. Adds apples to apple trees for admin apple pickers (#PR)
+#    issue: Add a Jira ticket or issue number, e.g. DP-12345 or 124
+#    impact: Enter a valid impact, e.g. Minor
+
+#___EXAMPLE___
+Changed:
+  - project: React, Patternlab
+    component: Key Message
```
Comment (clairesunstudio) [inline changelogs/DP-22263.yml]:
```suggestion
    component: KeyMessage
```
Please use PascalCase

File: changelogs/DP-22263.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    description: Adjusted to support reponsive images (picture)
+    issue: DP-22263
+    impact: Minor
+  - project: React, Patternlab
+    component: Marketing page
```
Comment (clairesunstudio) [inline changelogs/DP-22263.yml]:
```suggestion
    component: MarketingCampaign
```

File: changelogs/DP-22263.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+#    component: Enter one or multiple (comma separated) valid component name(s) - see ___COMPONENT___ e.g. Color
+#    description: Describe the change and add the PR number. see ___DESCRIPTION___ e.g. Adds apples to apple trees for admin apple pickers (#PR)
+#    issue: Add a Jira ticket or issue number, e.g. DP-12345 or 124
+#    impact: Enter a valid impact, e.g. Minor
+
+#___EXAMPLE___
+Changed:
+  - project: React, Patternlab
+    component: Key Message
+    description: Adjusted to support reponsive images (picture)
```
Comment (clairesunstudio) [inline changelogs/DP-22263.yml]:
```suggestion
    description: Adjusted to support responsive images (#1484)
```

**Comment by clairesunstudio** [comment]
@Javi-er, the order of the images is the opposite of the screen sizes. Not sure if this is just in mayflower? Are these changes just for mayflower or it will impact openness? 

I'm seeing the largest image used for the smallest breakpoint in mayflower 
Screen Shot 2021-09-13 at 11 27 14 AM

And the small image used on all screen sizes larger than 481px
Screen Shot 2021-09-13 at 11 28 13 AM

**Comment by clairesunstudio** [comment]
UAT links: 
Picture: https://mayflower.digital.mass.gov/patternlab/b/feature/DP-22263_drupal-responsive-images/?p=atoms-picture-element (**NOT** working)
~Image responsive: https://mayflower.digital.mass.gov/patternlab/b/feature/DP-22263_drupal-responsive-images/?p=atoms-image-responsive~
Key message: https://mayflower.digital.mass.gov/patternlab/b/feature/DP-22263_drupal-responsive-images/?p=organisms-key-message (working)
Key message (overlay): https://mayflower.digital.mass.gov/patternlab/b/feature/DP-22263_drupal-responsive-images/?p=organisms-key-message-as-solid-image-primary (working)
Marketing campaign: https://mayflower.digital.mass.gov/patternlab/b/feature/DP-22263_drupal-responsive-images/?p=pages-marketing-campaign (**NOT** working)

File: packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json [hunk 6]
```diff
@@ -8,7 +8,28 @@
         "level": "1"
       },
       "textOverlay": "c-gray-lightest",
-      "bgImage": "../../assets/images/placeholder/1920x1080-sign.jpg",
+      "bgImage": {
+        "path": "@atoms/09-media/picture-element.twig",
+        "data": {
+          "picture": {
+            "defaultImage": "../../assets/images/placeholder/1600x600.png",
+            "srcSets": [
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json:16]:
```suggestion
            "srcSets": [
```

the order of these images and breakpoints should be reversed

File: packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json [hunk 7]
```diff
@@ -1,18 +1,18 @@
...
+        "url": "../../assets/images/placeholder/900x700.png",
+        "width": "910"
       },
       {
-        "url": "../../assets/images/placeholder/600x450.png",
-        "width": "620"
+        "url": "../../assets/images/placeholder/1600x600.png",
+        "width": "1200"
       }
     ]
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json:17]: the order of these images and breakpoints should be reversed

File: packages/patternlab/styleguide/source/_patterns/03-organisms/by-template/key-message.twig [hunk 8]
```diff
@@ -55,9 +55,14 @@
...
-<section {% if id %}id="{{ id }}" {% endif %}class="ma__key-message{{ msgClass }}{% if bgImage %} ma__key-message--image{% endif %}{% if keyMessage.textOverlay and bgImage %} ma__key-message--image-overlay{% endif %}{% if positionTop %} ma__key-message--top{% endif %}"{% if bgImage %} style="background-image: url('{{ bgImage }}')"{% endif %}>
-  {% if bgImage and positionTop %}
-    <div id="img{{ id }}"  class="ma__key-message__inline-image" {% if keyMessage.mobileBgImage and positionTop %} style="background-image: url('{{ keyMessage.mobileBgImage }}')"{% endif %}>
+<section {% if id %}id="{{ id }}" {% endif %}class="ma__key-message{{ msgClass }}{% if bgImage %} ma__key-message--image{% endif %}{% if keyMessage.textOverlay and bgImage %} ma__key-message--image-overlay{% endif %}{% if positionTop %} ma__key-message--top{% endif %}">
+  {% if bgImage%}
+    <div id="img{{ id }}"  class="ma__key-message__inline-image">
+      {% if bgImage.path %}
+        {% include bgImage.path with bgImage.data %}
+      {% else %}
+        {{ bgImage }}
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/03-organisms/by-template/key-message.twig:66]: is this how the responsive `picture` element is passed in from the drupal responsive image module?  If so, can you add a comment in the check to explain why this is added for future reference? Thanks!

File: packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json (hunk 6 above)
Comment (Javi-er) [reply packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json:16]: done

File: packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json (hunk 7 above)
Comment (Javi-er) [reply packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json:17]: done

File: packages/patternlab/styleguide/source/_patterns/03-organisms/by-template/key-message.twig (hunk 8 above)
Comment (Javi-er) [reply packages/patternlab/styleguide/source/_patterns/03-organisms/by-template/key-message.twig:66]: I added comments in the template to document the purpose of those two lines.

**Comment by clairesunstudio** [comment]
Found another issue:
Screen Shot 2021-09-13 at 12 29 54 PM
The largest image is being loaded and rendered on the smallest size (< 480px)

See:
https://mayflower.digital.mass.gov/patternlab/b/feature/DP-22263_drupal-responsive-images/?p=atoms-picture-element
https://mayflower.digital.mass.gov/patternlab/b/feature/DP-22263_drupal-responsive-images/?p=pages-marketing-campaign

**Comment by Javi-er** [comment]
> Found another issue:
> Screen Shot 2021-09-13 at 12 29 54 PM
> The largest image is being loaded and rendered on the smallest size (< 480px)
> 
> See:
> https://mayflower.digital.mass.gov/patternlab/b/feature/DP-22263_drupal-responsive-images/?p=atoms-picture-element
> https://mayflower.digital.mass.gov/patternlab/b/feature/DP-22263_drupal-responsive-images/?p=pages-marketing-campaign

Just fixed it, the default image was wrong for mayflower.

**Comment by clairesunstudio** [comment]
ok setting the default images to be the smallest version, will mean that's the one image that will be loaded across all screen sizes on unsupported browsers - resulted in graininess on desktop. 
Screen Shot 2021-09-13 at 12 51 02 PM
But since the hero image is not core functionality, I think it should be okay

File: packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json [hunk 9]
```diff
@@ -1,18 +1,18 @@
 {
   "picture": {
-    "defaultImage": "../../assets/images/placeholder/400x300.png",
+    "defaultImage": "../../assets/images/placeholder/460x220.png",
     "srcSets": [
       {
-        "url": "../../assets/images/placeholder/900x700.png",
-        "width": "1050"
+        "url": "../../assets/images/placeholder/1600x600.png",
+        "width": "1200"
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json]:
```suggestion
        "width": "1201"
```

File: packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json [hunk 10]
```diff
@@ -1,18 +1,18 @@
...
       },
       {
-        "url": "../../assets/images/placeholder/800x450.png",
-        "width": "780"
+        "url": "../../assets/images/placeholder/900x700.png",
+        "width": "910"
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json]:
```suggestion
        "width": "911"
```

File: packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json [hunk 11]
```diff
@@ -1,18 +1,18 @@
...
       },
       {
-        "url": "../../assets/images/placeholder/600x450.png",
-        "width": "620"
+        "url": "../../assets/images/placeholder/500x300.png",
+        "width": "480"
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/01-atoms/09-media/picture-element.json]:
```suggestion
        "width": "481"
```

File: packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json [hunk 12]
```diff
@@ -8,7 +8,28 @@
...
-      "bgImage": "../../assets/images/placeholder/1920x1080-sign.jpg",
+      "bgImage": {
+        "path": "@atoms/09-media/picture-element.twig",
+        "data": {
+          "picture": {
+            "defaultImage": "../../assets/images/placeholder/460x220.png",
+            "srcSets": [
+              {
+                "url": "../../assets/images/placeholder/1600x600.png",
+                "width": "1200"
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json]:
```suggestion
                "width": "1201"
```

File: packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json [hunk 13]
```diff
@@ -8,7 +8,28 @@
...
+              },
+              {
+                "url": "../../assets/images/placeholder/900x700.png",
+                "width": "910"
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json]:
```suggestion
                "width": "911"
```

File: packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json [hunk 14]
```diff
@@ -8,7 +8,28 @@
...
+              },
+              {
+                "url": "../../assets/images/placeholder/500x300.png",
+                "width": "480"
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/05-pages/marketing-campaign.json]:
```suggestion
                "width": "481"
```

**Review by clairesunstudio — COMMENTED** [review]
Can you commit these suggestions, and fix failing backstop tests in patternlab? Otherwise looks good to me!

**Comment by Javi-er** [comment]
> ok setting the default images to be the smallest version, will mean that's the one image that will be loaded across all screen sizes on unsupported browsers - resulted in graininess on desktop.
> Screen Shot 2021-09-13 at 12 51 02 PM
> But since the hero image is not core functionality, I think it should be okay

Right, in this case it's not affected by that since Mayflower it's using  [Picturefill](https://github.com/scottjehl/picturefill) , I found it  because took a look on IE11 and it's looking fine, see:
j-r-capture 2021-09-13 a la(s) 14 18 23


https://github.com/massgov/mayflower/blob/develop/packages/patternlab/styleguide/package.json#:~:text=%22-,picturefill,-%22%3A%20%22%5E3.0.3%22%2C

**Comment by Javi-er** [comment]
> Can you commit these suggestions, and fix failing backstop tests in patternlab? Otherwise looks good to me!

just commited those, and I'll take a look to Backstop in my local.

**Comment by Javi-er** [comment]
@clairesunstudio Backstop tests are now approved.

**Review by clairesunstudio — APPROVED** [review]

**Pull request merged**